src/basic/user-util.c

```diff
@@ -602,3 +604,27 @@ bool valid_home(const char *p) {
 
         return true;
 }
+
+int maybe_setgroups(size_t size, const gid_t *list) {
+        static int cached_can_setgroups = -1;
+        if (cached_can_setgroups < 0) {
+                _cleanup_free_ char *setgroups_content = NULL;
+                int r;
+
+                r = read_one_line_file("/proc/self/setgroups", &setgroups_content);
+                if (r < 0 && errno != ENOENT)
+                        return r;
+                if (r < 0) {
+                        /* old kernels don't have /proc/self/setgroups, so assume we can use it */
+                        cached_can_setgroups = true;
+                } else {
+                        cached_can_setgroups = !!strcmp(setgroups_content, "deny");
+                        if (!cached_can_setgroups)
+                                log_debug("skip setgroups, /proc/self/setgroups is set to 'deny'");
+                }
+        }
+        if (!cached_can_setgroups)
+                return 0;
+
+        return setgroups(size, list);
+}
```

Comment:
I think one further refinement for this call would make sense: only skip the setgroups() invocation if the groups list is reset (i.e. size == 0), and continue generate an error if the list is anything but empty. This way if people actually set SupplementaryGroups= to something non-empty, then we'll continue to fail (and I think we should), but for the typical case where the list is just supposed to be reset we eat up the error. Does that make sense?